sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +385,464 @@ def transform(self, X):
+
+        # pre-processing of X
+        X = check_array(X, dtype=np.float64, force_all_finite=False)
+        X = np.asarray(X, order="F")
+
+        # parse min and max values
+        self.min_value_ = np.nan if self.min_value is None else self.min_value
+        self.max_value_ = np.nan if self.max_value is None else self.max_value
+
+        # initial imputation
+        mask_missing_values = _get_mask(X, self.missing_values)
+        self.initial_imputer_ = Imputer(missing_values=self.missing_values,
+                                        strategy=self.initial_fill_method,
+                                        axis=0)
+        X_filled = self.initial_imputer_.fit_transform(X)
+        self._val_inds = self.initial_imputer_._valid_statistics_inds
+        X = X[:, self._val_inds]
+        mask_missing_values = mask_missing_values[:, self._val_inds]
+
+        # perform imputations
+        self._trained_estimator_triplets = []
+        n_samples, n_features = X_filled.shape
+        total_rounds = self.n_burn_in + self.n_imputations
+        results_list = []
+        if self.verbose:
+            print("[MICE] Completing matrix with shape %s" % (X.shape,))
+            start_t = time()
+            mice_msg = '[MICE] Ending imputation round '
+        for m in range(total_rounds):
+            # order in which to impute
+            ordered_inds = self._get_ordered_inds(mask_missing_values)
+
+            # abs_correlation matrix is used to choose a subset of other
+            # features to impute from  f
+            abs_corr_mat = self._get_abs_correlation_matrix(X_filled)
+
+            # Fill in each feature in the order of ordered_inds
+            for this_feat_ind in ordered_inds:
+                i = this_feat_ind  # the tyranny of pep8
+                other_feat_inds = self._get_other_feat_inds(n_features,
+                                                            i,
+                                                            abs_corr_mat)
+                X_filled, estimator = self._impute_one_feature(
+                    X_filled,
+                    mask_missing_values,
+                    this_feat_ind,
+                    other_feat_inds
+                )
+                estimator_triplet = (this_feat_ind,
+                                     other_feat_inds,
+                                     estimator)
+                self._trained_estimator_triplets.append(estimator_triplet)
+
+            if m >= self.n_burn_in:
+                results_list.append(X_filled[mask_missing_values])
+            if self.verbose:
+                print(mice_msg + 'round %d/%d, elapsed time %0.2f'
+                      % (m + 1, total_rounds, time() - start_t))
+
+        if len(results_list) > 0:
+            X[mask_missing_values] = np.array(results_list).mean(axis=0)
+        else:
+            X[mask_missing_values] = X_filled[mask_missing_values]
+
+        return X
+
+    def transform(self, X):
```

Comment:
why fit_transform is not calling fit + transform. What I mean is can we reduce the code such that fit_transform does not repeat some part of transform

Reply:
In short: to save time. When you do `fit`, you get `transform` for free basically. I can restructure, but it'll just be slower.

Reply:
using a private function, it should be possible to avoid code duplication without losing so much speed.

Reply:
Ok, I'll give it a try.

Reply:
So is it sufficient to have a private `_fit_transform`? And it's called by both `fit` and `fit_transform`?

Reply:
What I mean is to create some private function where to put the code in common in `fit_transform` and `transform`

Reply:
They do quite different things in this case. This kind of refactoring would be unnatural. The reason being is that both the fitting and the transformation happens one imputation at a time. I don't really see which parts of `fit` and `transform` are "common". The common part is `_impute_one_feature`, which is already its own function.